Add new line to be specified for JSON formatting

martincostello opened the pull request:
Allow the new line string to use for indented JSON to be specified through options.

Resolves #84117.

martincostello committed:
Add new line to be specified for JSON formatting

Allow the new line string to use for indented JSON to be specified through options.
Resolves #84117.

martincostello commented:
@eiriktsarpalis I've left this in draft for now before I dig around tomorrow and see if I've missed anything, but feel free to start reviewing at your convenience.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/Resources/Strings.resx:
```diff
@@ -717,4 +717,7 @@
     <value>Indentation size must be between {0} and {1}.</value>
   </data>
+  <data name="InvalidNewLine" xml:space="preserve">
+    <value>Only the strings '\n' or '\r\n' are permitted.</value>
```
It's kind of odd to see single quotes delimit string literals, but OTOH I've never seen double quotes used in exception messages either.

martincostello replied:
Yeah I wasn't quite sure what exactly to use here, but many of the other messages in this file quote with single quotes. Happy to reword if there's a better way to express it.

eiriktsarpalis replied:
@bartonjs might have opinions on this

bartonjs replied:
While we often use single quotes for things other than System.Char, it generally seems to always be for things other than strings.  (Often chars (from parsers), occasionally enum values, or property names, etc).

When it comes to actual strings, though, one can find examples both ways.  To me this feels closest to `guid.ToString(format)` rejecting anything except a few limited inputs... which uses double-quotes:

https://github.com/dotnet/runtime/blob/839c642c3abebe5ba7fcd18b5b03dee50dc1b76e/src/libraries/System.Private.CoreLib/src/Resources/Strings.resx#L2401-L2406

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/JsonConstants.cs:
```diff
@@ -28,6 +28,9 @@ internal static partial class JsonConstants
         public const byte UtcOffsetToken = (byte)'Z';
         public const byte TimePrefix = (byte)'T';
 
+        public const string NewLineLineFeed = "\n";
+        public const string NewLineCarriageReturnLineFeed = "\r\n";
```
Nit: kind of redundant naming?
```suggestion
        public const string LineFeed = "\n";
        public const string CarriageReturnLineFeed = "\r\n";
```

martincostello replied:
LineFeed is already taken on line 14 as a byte constant.

stephentoub replied:
Do we even need these to be named?

martincostello replied:
I was just following the existing pattern based off of the `TabIndentCharacter` and `DefaultIndentCharacter` constants and their usage.

I can just inline them if preferred, both of these new constants are used in two places each.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:
```diff
@@ -89,6 +89,7 @@ public static JsonSerializerOptions Web
         private bool _ignoreReadOnlyProperties;
         private bool _ignoreReadonlyFields;
         private bool _includeFields;
+        private string _newLine = Environment.NewLine;
```
Nit: maybe make initialization lazy on the public property itself?

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs:
```diff
@@ -141,6 +142,7 @@ public JsonSerializerOptions(JsonSerializerOptions options)
             _ignoreReadOnlyProperties = options._ignoreReadOnlyProperties;
             _ignoreReadonlyFields = options._ignoreReadonlyFields;
             _includeFields = options._includeFields;
+            _newLine = options._newLine;
```
The new field must also be taken into account in the `JsonSerializerOptions.Caching.cs` methods.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs:
```diff
@@ -150,6 +151,7 @@ private void SetOptions(JsonWriterOptions options)
             _options = options;
             _indentByte = (byte)_options.IndentCharacter;
             _indentLength = options.IndentSize;
+            _newLineLength = options.NewLine.Length;
```
Shouldn't this be set by the constructors?

eiriktsarpalis replied:
Can you add a Debug.Assert checking the value of `NewLine`?

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs:
```diff
@@ -61,6 +61,7 @@ public sealed partial class Utf8JsonWriter : IDisposable, IAsyncDisposable
         // Cache indentation settings from JsonWriterOptions to avoid recomputing them in the hot path.
         private byte _indentByte;
         private int _indentLength;
+        private int _newLineLength = s_defaultNewLineLength;
```
Shouldn't this be set by the constructors?

eiriktsarpalis replied:
Can you add a comment clarifying that 1 denotes LF and 2 denotes CRLF?

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs:
```diff
@@ -1024,7 +1026,7 @@ private void WriteEndIndented(byte token)
         private void WriteNewLine(Span<byte> output)
         {
             // Write '\r\n' OR '\n', depending on OS
```
Might need to update the comment as well.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs:
```diff
@@ -35,7 +35,7 @@ namespace System.Text.Json
     {
         // Depending on OS, either '\r\n' OR '\n'
-        private static readonly int s_newLineLength = Environment.NewLine.Length;
+        private static readonly int s_defaultNewLineLength = Environment.NewLine.Length;
```
If this is only used to specify the default value, perhaps it can be removed?

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonWriterOptionsTests.cs:
```diff
@@ -157,5 +168,17 @@ public static void JsonWriterOptions_IndentSize_OutOfRange(int size)
         }
+
+        [Theory]
+        [InlineData("")]
```
Could you add a few non-whitespace inputs for good measure?

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/OptionsTests.cs:
```diff
@@ -1416,7 +1416,14 @@ private static JsonSerializerOptions GetFullyPopulatedOptionsInstance()
                 else if (propertyType == typeof(string))
                 {
-                    property.SetValue(options, "\t");
+                    if (property.Name is "NewLine")
```
Nit: `nameof`?

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs:
```diff
@@ -36,6 +36,13 @@ public static void WriteIndentation(Span<byte> buffer, int indent, byte indentBy
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void ValidateNewLine(string value)
+        {
+            if (value is not JsonConstants.NewLineLineFeed and not JsonConstants.NewLineCarriageReturnLineFeed)
```
If someone wants to explicitly set the default behavior, how do they do that? Shouldn't null be allowed as well to indicate default?

eiriktsarpalis replied:
Setting to `Environment.NewLine` would work. I don't have a strong opinion, but there is precedent in returning the default setting in the other properties of JSO.

martincostello commented:
Failing test is [`EnsureConsoleLoggerOptions_ConfigureOptions_SupportsAllProperties()`](https://github.com/dotnet/runtime/blob/f0991722db63e0b4962439469efb4ea1b72923f3/src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerConfigureOptions.cs#L16-L29) - I need to update the JSON console logger options to account for the new property.

martincostello committed:
Address review feedback

- Cater for `_newLine` in JsonSerializerOptions caching.
- Lazily initialize field.
- Allow null to reset to default.
- Add assertions.
- Add/update comments.
- Use `nameof()`.
- Remove redundant field.
- Extend tests.

martincostello committed:
Update Logging.Console tests

- Update property count to fix assertion.
- Update test to validate `NewLine` can be set/bound.

martincostello committed:
Only normalize line endings if needed

Only normalize the line endings if the `JsonWriterOptions` are not using the defaults.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs:
```diff
@@ -36,6 +36,13 @@ public static void WriteIndentation(Span<byte> buffer, int indent, byte indentBy
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Why does this need to be explicitly inlined? It doesn't appear to be called in any of the hot paths.

eiriktsarpalis replied:
On closer inspection, I see that this just replicates other validation methods in the same file. Seems questionable, but keeping consistency would be preferable.

martincostello replied:
I just copied what all the other validation methods were already doing.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -561,6 +562,7 @@ public int GetHashCode(JsonSerializerOptions options)
                 AddHashCode(ref hc, options._unmappedMemberHandling);
                 AddHashCode(ref hc, options._defaultBufferSize);
                 AddHashCode(ref hc, options._maxDepth);
+                AddHashCode(ref hc, options._newLine ?? Environment.NewLine); // Null is equivalent to the default
```
You could try calling the property directly to avoid replicating the default value here.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -504,6 +504,7 @@ public bool Equals(JsonSerializerOptions? left, JsonSerializerOptions? right)
                     left._unmappedMemberHandling == right._unmappedMemberHandling &&
                     left._defaultBufferSize == right._defaultBufferSize &&
                     left._maxDepth == right._maxDepth &&
+                    left._newLine == right._newLine &&
```
Similarly, this should call the property because comparison could result in false negatives if either property has not been initialized.

martincostello commented:
I'm not sure why the Mono tests are failing - there's something in the logs about how some of the tests take a long time to run under the interpreter, so I wonder if the extra test cases I've added to the combinations for `NewLine` have tipped it over the edge for the 1 hour timeout?

martincostello committed:
Address feedback

- Access lazily initialized field through property.
- Update hash code assertion.

martincostello committed:
Update exception message

Use similar format string to `Format_InvalidGuidFormatSpecification`/

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -591,13 +593,13 @@ static void AddListHashCode<TValue>(ref HashCode hc, ConfigurationList<TValue>?
                 static void AddHashCode<TValue>(ref HashCode hc, TValue? value)
                 {
-                    if (typeof(TValue).IsValueType)
+                    if (typeof(TValue).IsValueType || !typeof(TValue).IsSealed)
```
`IsValue` implies `IsSealed` so the first check is redundant.
```suggestion
                    if (!typeof(TValue).IsSealed)
```

martincostello replied:
That makes `JsonSerializerOptions_ReuseConverterCaches()` fail - I'm not sure exactly why right now as I just installed the preview 3 SDK and now Visual Studio won't let me debug the tests for some reason I've just to fix.

eiriktsarpalis replied:
My bad, this should have simply been `if (typeof(TValue).IsValue)` without the negation.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -591,13 +593,13 @@ static void AddListHashCode<TValue>(ref HashCode hc, ConfigurationList<TValue>?
                     else
                     {
-                        Debug.Assert(!typeof(TValue).IsSealed, "Sealed reference types like string should not use this method.");
+                        // Unsealed types do not use this branch to avoid user-defined equality from being used.
```
```suggestion
                        // Use the built-in hashcode for types that could be overriding GetHashCode().
```

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterOptions.cs:
```diff
@@ -135,11 +137,33 @@ public bool SkipValidation
+            set
+            {
+                JsonWriterHelper.ValidateNewLine(value);
+                if (value is not null && value != Environment.NewLine)
```
Can `value` be null after validation is done?

martincostello replied:
I added guards into the two callers to handle the null case.

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/OptionsTests.cs:
```diff
@@ -516,6 +516,60 @@ public static void MaxDepthRead()
+        {
+            var options = new JsonSerializerOptions();
+            options.NewLine = "\n";
+            options.NewLine = null!;
```
I find that behavior surprising. Assuming we want to use `null` to denote default that should be reflected in the nullability annotation for the property setter.

martincostello replied:
Yeah I thought it a bit weird too, but annotations or not there was nothing to stop a user passing in null anyway.

Either the properties need changing to `string?`, or I can revert back to what I had originally and force the user to specify `Environment.NewLine` if they want to go back to the default.

eiriktsarpalis replied:
General practice is to throw argument null exception if an input parameter is non-nullable. I think using `Environment.NewLine` as the default works and results in a simpler API contract.

martincostello replied:
I've updated things to do the normal "throw if null` behaviour, so the user would need to explicitly set to `Environment.NewLine` to restore the default behaviour.

eiriktsarpalis replied:
Thanks. We can always move to allow nulls in the future without introducing breaking changes.

martincostello commented:
> I'm not sure why the Mono tests are failing

Looks like these are [known issues](https://github.com/dotnet/runtime/pull/100890/checks?check_run_id=23716492244).

martincostello committed:
Address feedback

Reword comment as suggested.

martincostello committed:
Address feedback

- Simplify condition.
- Disallow null for `string NewLine` properties.

eiriktsarpalis commented:
> > I'm not sure why the Mono tests are failing
> 
> Looks like these are [known issues](https://github.com/dotnet/runtime/pull/100890/checks?check_run_id=23716492244).

We generally use the "Build Analysis" leg to determine if there any test failures that haven't been flagged as known issues.

eiriktsarpalis reviewed (approved):
Thanks!
